File: backend/app/models/schemas.py
"""Request/response schemas for the Phase 1 API surface."""

from __future__ import annotations

from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field


PageType = Literal[
    "topic_hub",
    "combinatorial",
    "sacred_text_chapter",
    "sacred_text_shloka",
    "sanskrit_lexicon",
]


class GenerateRequest(BaseModel):
    topic: str = Field(min_length=3)
    page_type: PageType
    goal: Optional[str] = None
    audience: Optional[str] = "spiritual seekers"


class QualityDimensionScore(BaseModel):
    score: float
    weight: float
    notes: str


class QualityScorecard(BaseModel):
    total_score: float
    passed: bool
    dimensions: Dict[str, QualityDimensionScore]
    violations: List[str] = Field(default_factory=list)


class GenerateResponse(BaseModel):
    topic: str
    page_type: PageType
    generated_content: str
    scorecard: QualityScorecard
    created_at: datetime = Field(default_factory=datetime.utcnow)


class TechniqueItem(BaseModel):
    source: str
    type: str
    text: str


class TechniquesResponse(BaseModel):
    techniques: List[TechniqueItem]


# ---------------------------------------------------------------------------
# Phase 2 Pipeline schemas
# ---------------------------------------------------------------------------

PipelineStatusLiteral = Literal[
    "queued",
    "researching",
    "research_review",
    "writing",
    "draft_review",
    "editing",
    "edit_review",
    "polishing",
    "final_review",
    "needs_review",
    "approved",
    "rejected",
    "failed",
]

PipelineFeedbackStageLiteral = Literal[
    "research_review",
    "draft_review",
    "edit_review",
    "final_review",
    "final",
]


class PipelineCreateRequest(BaseModel):
    topic: str = Field(min_length=3)
    page_type: PageType
    description: Optional[str] = None
    reference_links: List[str] = Field(default_factory=list)
    key_angles: List[str] = Field(default_factory=list)
    source_notes: Optional[str] = None
    goal: Optional[str] = None
    audience: Optional[str] = "spiritual seekers"
    quality_threshold: Optional[float] = Field(default=None, ge=1.0, le=10.0)
    revision_limit: Optional[int] = Field(default=None, ge=0, le=5)


class ContentOutputSchema(BaseModel):
    id: str
    version: int
    stage: str
    agent: str
    content: str
    revision_notes: Optional[str] = None
    scorecard_json: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class FeedbackEntrySchema(BaseModel):
    id: str
    stage: str
    action: str
    diff: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class PipelineStatus(BaseModel):
    """Full pipeline detail including all outputs and feedback."""
    id: str
    topic: str
    page_type: str
    description: Optional[str] = None
    reference_links: List[str] = Field(default_factory=list)
    key_angles: List[str] = Field(default_factory=list)
    source_notes: Optional[str] = None
    goal: Optional[str] = None
    audience: Optional[str] = None
    context_module: Optional[str] = None
    status: str
    revision_count: int
    quality_threshold: float
    revision_limit: int
    final_score: Optional[float] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    outputs: List[ContentOutputSchema] = Field(default_factory=list)
    feedback: List[FeedbackEntrySchema] = Field(default_factory=list)

    class Config:
        from_attributes = True


class PipelineListItem(BaseModel):
    """Lightweight card data for the Kanban board."""
    id: str
    topic: str
    page_type: str
    status: str
    revision_count: int
    final_score: Optional[float] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PipelineListResponse(BaseModel):
    pipelines: List[PipelineListItem]
    total: int


class ApproveRequest(BaseModel):
    notes: Optional[str] = None


class AdvanceRequest(BaseModel):
    notes: Optional[str] = None


class RejectRequest(BaseModel):
    notes: str


class ReviseRequest(BaseModel):
    notes: str
    target_dimensions: Optional[List[str]] = None


class FeedbackRequest(BaseModel):
    stage: PipelineFeedbackStageLiteral
    action: Literal["approve", "reject", "edit"]
    diff: Optional[str] = None
    notes: Optional[str] = None

File: src/lib/pipelines/schemas.ts
/**
 * Zod schemas for pipeline backend responses.
 *
 * These enforce a contract at the proxy boundary between the Python backend
 * and the Next.js frontend. If the backend changes its output format, these
 * schemas will throw a clear ZodError instead of silently ingesting bad data.
 */

import { z } from "zod";

// --- Shared enums ---

export const PipelinePageTypeSchema = z.enum([
    "topic_hub",
    "combinatorial",
    "sacred_text_chapter",
    "sacred_text_shloka",
    "sanskrit_lexicon",
]);

export const PipelineStatusSchema = z.enum([
    "queued",
    "researching",
    "research_review",
    "writing",
    "draft_review",
    "editing",
    "edit_review",
    "polishing",
    "final_review",
    "needs_review",
    "approved",
    "rejected",
    "failed",
]);

export const PipelineFeedbackActionSchema = z.enum(["approve", "reject", "edit"]);

// --- Output schemas ---

export const PipelineOutputSchema = z.object({
    id: z.string(),
    version: z.number(),
    stage: z.string(), // Intentionally loose — backend may add new stages
    agent: z.string(),
    content: z.string(),
    revision_notes: z.string().nullish(),
    scorecard_json: z.string().nullish(),
    created_at: z.string(),
});

export const PipelineFeedbackEntrySchema = z.object({
    id: z.string(),
    stage: z.string(),
    action: PipelineFeedbackActionSchema,
    diff: z.string().nullish(),
    notes: z.string().nullish(),
    created_at: z.string(),
});

// --- List item schema ---

export const PipelineListItemSchema = z.object({
    id: z.string(),
    topic: z.string(),
    page_type: z.string(), // Loose — may include new types from backend
    status: z.string(),
    revision_count: z.number(),
    final_score: z.number().nullish(),
    created_at: z.string(),
    updated_at: z.string(),
});

// --- Detail schema (used by materialize route) ---

export const PipelineDetailSchema = PipelineListItemSchema.extend({
    description: z.string().nullish(),
    reference_links: z.array(z.string()),
    key_angles: z.array(z.string()),
    source_notes: z.string().nullish(),
    goal: z.string().nullish(),
    audience: z.string().nullish(),
    context_module: z.string().nullish(),
    quality_threshold: z.number(),
    revision_limit: z.number(),
    error_message: z.string().nullish(),
    outputs: z.array(PipelineOutputSchema),
    feedback: z.array(PipelineFeedbackEntrySchema),
});

// --- List response schema ---

export const PipelineListResponseSchema = z.object({
    pipelines: z.array(PipelineListItemSchema),
    total: z.number(),
});

// --- Materialization contract ---
// What the materialize endpoint expects from a validated pipeline detail

export const MaterializationContractSchema = z.object({
    id: z.string().min(1),
    topic: z.string().min(1),
    status: z.literal("approved"),
    page_type: z.string(),
    description: z.string().nullish(),
    reference_links: z.array(z.string()),
    key_angles: z.array(z.string()),
    source_notes: z.string().nullish(),
    goal: z.string().nullish(),
    audience: z.string().nullish(),
    context_module: z.string().nullish(),
    quality_threshold: z.number(),
    final_score: z.number().nullish(),
    error_message: z.string().nullish(),
    outputs: z.array(PipelineOutputSchema).min(1, "Pipeline must have at least one output to materialize"),
});

// --- Type exports inferred from schemas ---

export type ValidatedPipelineDetail = z.infer<typeof PipelineDetailSchema>;
export type ValidatedPipelineListResponse = z.infer<typeof PipelineListResponseSchema>;
export type ValidatedMaterializationContract = z.infer<typeof MaterializationContractSchema>;

export function formatSchemaValidationError(error: z.ZodError): string {
    return error.issues.map((issue) => `${issue.path.join(".") || "root"}: ${issue.message}`).join("; ");
}

// --- Validation helpers ---

/**
 * Validate a pipeline detail response from the backend.
 * Throws ZodError with clear field-level messages on mismatch.
 */
export function validatePipelineDetail(data: unknown): ValidatedPipelineDetail {
    return PipelineDetailSchema.parse(data);
}

/**
 * Validate a pipeline list response from the backend.
 */
export function validatePipelineListResponse(data: unknown): ValidatedPipelineListResponse {
    return PipelineListResponseSchema.parse(data);
}

export function validateMaterializationContract(data: unknown): ValidatedMaterializationContract {
    return MaterializationContractSchema.parse(data);
}

/**
 * Safe validation — returns { success, data, error } instead of throwing.
 */
export function safeParsePipelineDetail(data: unknown) {
    return PipelineDetailSchema.safeParse(data);
}
